### PR TITLE
added tests for bounding functions & utilities

### PR DESCRIPTION
**Pull request by Tominapa**
## Fixes/Addresses/Summary/Motivation:

Adds bounding functions and utilities. Tests implemented for all.

Note for future iterations of the bounding utility:

- Bounding for v_F_DiscreteTrucked requires changes to strategic model to be implemented: https://github.com/project-pareto/project-pareto/pull/254#discussion_r1260083135
- Bounding for v_Q  (p_Q_UB) is working, but can be improved: https://github.com/project-pareto/project-pareto/pull/254#discussion_r1255832089
- Remove redundant free() and unfix() references

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my
contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.md file
   at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has
   rights to intellectual property that includes these contributions, I represent that I have
   received permission to make contributions and grant the required license on behalf of that
   employer.

### CODE REVIEW

**Comment by lbianchi-lbl** [comment]
This is also ready to merge, pending reviews.

File: pareto/utilities/bounding_functions.py
Comment (melody-shellman) [inline pareto/utilities/bounding_functions.py:35]: Noticed this TODO. Is it still something that is needed?

File: pareto/utilities/model_modifications.py
Comment (melody-shellman) [inline pareto/utilities/model_modifications.py:1]: These are nifty and I could see being quite useful!

File: pareto/utilities/bounding_functions.py
Comment (melody-shellman) [inline pareto/utilities/bounding_functions.py:114]: What does the "NOTE" here mean?

File: pareto/utilities/bounding_functions.py
Comment (melody-shellman) [inline pareto/utilities/bounding_functions.py:13]: I think it would be helpful to add a note here that this method adds bounds to the following variables: v_F_Piped, v_F_Trucked, quality.v_Q, and v_F_DiscretePiped with a note on why others variables are not included.

**Review by melody-shellman — COMMENTED** [review]
Looking good! I have a few comments/suggestions proposed.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:35]: This will require a change in the corresponding PARETO parameter in the strategic model code. I built this in for future reference. At this point, I would rather keep this as-is and raise this in a future update.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:114]: I tested the commented line that follows on 115 where the bound is derived from the p_Q_UB parameter. The logic that is used to create that parameter is not correct, resulting in a bound that is too tight, and the model solution will change with that bound enforced. I kept it there for two reasons: one, so it is available for testing should the same idea be tried in future, and two, as a record that the approach did not function properly. The current bound is fairly loose, and there is room for improvement in future versions.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:13]: We could absolutely include other variables in future versions. The objective was to develop bounds for these three, being the variables involved in nonlinear functions in the MINLP formulation of the strategic model.

File: pareto/utilities/bounding_functions.py
Comment (tarnold17) [reply pareto/utilities/bounding_functions.py:114]: @Tominapa that seems reasonable to me, but I think you might as well comment out all of the code for `model.quality.p_Q_UB` (lines 101-109) for now, since it is currently unused.

File: pareto/utilities/model_modifications.py
Comment (tarnold17) [inline pareto/utilities/model_modifications.py:60]:
```suggestion
```
Minor, but you can remove this line if you want. Python returns `None` by default if a `return` statement is not explicitly included.

File: pareto/utilities/model_modifications.py
Comment (tarnold17) [inline pareto/utilities/model_modifications.py:46]:
```suggestion
```
Same comment here.

File: pareto/utilities/model_modifications.py
Comment (tarnold17) [inline pareto/utilities/model_modifications.py:44]: Why are you setting the lower bound to zero here instead of `None`? Do we have any variables that could conceivably take on negative values?

File: pareto/utilities/model_modifications.py
Comment (tarnold17) [inline pareto/utilities/model_modifications.py:41]: What does the `free()` function do?

File: pareto/tests/test_utilities.py
Comment (tarnold17) [inline pareto/tests/test_utilities.py:33]:
```suggestion
```
These imports are all unused and can be removed.

File: pareto/tests/test_utilities.py
Comment (tarnold17) [inline pareto/tests/test_utilities.py:48]: Two things:

1) Is there a reason you aren't using `@pytest.fixture(scope="module")` like we do in `test_strategic_model.py`?
2) This function doesn't just build the strategic model, it also solves it. So I recommend changing the name of the function to reflect that fact.

**Review by tarnold17 — COMMENTED** [review]
I don't see any problems, but I do have quite a few questions.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:39]: I did this in case any other options become available in future. This is more verbose, but also more future-proof.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:67]: Solvers treat bounds and constraints differently. For MINLP formulations, having explicit bounds will be helpful.

Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:67]: Even if they are redundant.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:121]: I built this function to handle parameters that might not be defined on every index. At present, it isn't used, but it was used in earlier versions that had different logic.

File: pareto/utilities/model_modifications.py
Comment (Tominapa) [reply pareto/utilities/model_modifications.py:15]: Probably. It was probably copied in with other code.

File: pareto/utilities/bounding_functions.py
Comment (tarnold17) [reply pareto/utilities/bounding_functions.py:67]: > Solvers treat bounds and constraints differently.

I did not realize this fact. Thanks for explaining!

File: pareto/utilities/bounding_functions.py
Comment (tarnold17) [reply pareto/utilities/bounding_functions.py:121]: @Tominapa Do you expect it will be needed later? If we have no reason to believe that this function will be needed later, then I'd say just get rid of it.

File: pareto/tests/test_utilities.py
Comment (Tominapa) [reply pareto/tests/test_utilities.py:33]: Addressed

File: pareto/tests/test_utilities.py
Comment (Tominapa) [reply pareto/tests/test_utilities.py:48]: 1. I don't know what that first part is for. It can't be important though, or someone would have mentioned it earlier.
2. I need the strategic model with its quality block. If I could get it *without* solving the model, I would. The naming serves its purpose, in that respect.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:13]: Added

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:35]: The parameter p_max_number_of_trucks will need to be available within the primary model block, or else this bound can only be added to specific subtypes. Moreover, the units on the parameter are [unitless] but I have a strong argument for making the units for p_max_number_of_trucks [week^-1]. I think this will need to be a follow-up PR once this first version is merged.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:114]: Commented

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:137]: Corrected

File: pareto/utilities/model_modifications.py
Comment (Tominapa) [reply pareto/utilities/model_modifications.py:15]: Removed.

File: pareto/utilities/model_modifications.py
Comment (Tominapa) [reply pareto/utilities/model_modifications.py:41]: The free() method unfixes the fixed value of a variable.

File: pareto/utilities/model_modifications.py
Comment (Tominapa) [reply pareto/utilities/model_modifications.py:46]: I like the explicit "return None" It's readable.

File: pareto/utilities/bounding_functions.py
Comment (nareshsusarla) [inline pareto/utilities/bounding_functions.py:71]: What does it mean "using bounds rule would require redefining the variable"

File: pareto/utilities/model_modifications.py
Comment (nareshsusarla) [inline pareto/utilities/model_modifications.py:23]: I suppose, if you use `component_data_objects` instead of `component_objects`, you need not check if the var object is indexed and then it could simplify your code. Unless you need it for another reason

File: pareto/utilities/model_modifications.py
Comment (nareshsusarla) [inline pareto/utilities/model_modifications.py:38]: it may be better to use `is_binary()` method here

File: pareto/utilities/model_modifications.py
Comment (nareshsusarla) [reply pareto/utilities/model_modifications.py:40]: why both `unfix()` and `free()`, they are same methods?

File: pareto/utilities/bounding_functions.py
Comment (nareshsusarla) [reply pareto/utilities/bounding_functions.py:130]: i  would think its best to remove it if the logic has changed

**Review by nareshsusarla — CHANGES_REQUESTED** [review]

File: pareto/utilities/bounding_functions.py
Comment (nareshsusarla) [inline pareto/utilities/bounding_functions.py:58]: **If the pipeline** connecting any two arcs **is bi-directional** and the model decides to install pipelines in both directions (this may be likely for trunk lines in new development areas), the choice of bound could be limiting the flow (at least conceptually)

**Review by nareshsusarla — CHANGES_REQUESTED** [review]

File: pareto/utilities/model_modifications.py
Comment (tarnold17) [reply pareto/utilities/model_modifications.py:45]: Yep, it's a matter of style I guess.

File: pareto/utilities/model_modifications.py
Comment (tarnold17) [reply pareto/utilities/model_modifications.py:40]: I would also like an answer to @nareshsusarla's question here.

File: pareto/tests/test_utilities.py
Comment (tarnold17) [inline pareto/tests/test_utilities.py:37]:
```suggestion
```
This import is unused, so these lines can be removed.

File: pareto/tests/test_utilities.py
Comment (tarnold17) [reply pareto/tests/test_utilities.py:41]: @Tominapa your reasoning on 1 is fallacious. I was the second person to review this PR - you can't dismiss my question as unimportant simply because the first reviewer didn't mention it in their review. That said, while you could use a pytest fixture here, I don't think you need to - I was just wondering if there was a reason why you had chosen not to. You can learn more about fixtures [here](https://docs.pytest.org/en/6.2.x/fixture.html) and you can look at some of our other unit tests for examples of when/how to use them.

On 2 - as it is currently written, the function `build_strategic_model` allows you to build and solve the strategic model with whatever option you want for quality (as specified in the `config_dict` argument that the function takes). I understand why you need to solve the model in this function, so I suggest changing its name to `build_and_solve_strategic_model` to more accurately describe its behavior. In other parts of our codebase, functions that say "build" in the name generally just build a model without solving. So since this function both builds and also solves the model, you can prevent some potential confusion for other developers by changing the name to reflect that. This is admittedly a somewhat minor thing, but it's a change that will help anyone who reads this code somewhere down the line.

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:58]: To my knowledge there is not a "pipe" index that I could use to resolve this problem. That's a PARETO-level limitation as I see it. It won't be solved by these utility functions. I'm marking it as resolved without further consideration.

File: pareto/utilities/model_modifications.py
Comment (Tominapa) [reply pareto/utilities/model_modifications.py:38]: Perhaps it would.

File: pareto/utilities/model_modifications.py
Comment (Tominapa) [reply pareto/utilities/model_modifications.py:23]: Perhaps this is so.

File: pareto/utilities/model_modifications.py
Comment (Tominapa) [reply pareto/utilities/model_modifications.py:40]: The documentation for both is the same. In this sense, I see no harm in covering my bases. Perhaps a better question is this: Why does Pyomo have two identical methods?

File: pareto/utilities/bounding_functions.py
Comment (Tominapa) [reply pareto/utilities/bounding_functions.py:130]: I might need it later if we are going to incorporate more variables, or change the logic for quality variables. This code is by no means complete at this stage.

File: pareto/tests/test_utilities.py
Comment (Tominapa) [reply pareto/tests/test_utilities.py:41]: 1. Thanks for the pointer to the docs. I figured out why you are asking for the change based on the context there. It should be the same with the change, benefiting from any marginal test time reductions.
2. I see your point, but the key from my point of view is that I do not _want_ to solve the model at this stage. If I could construct the quality model without a solve call, I would. This would require a change within PARETO itself, which I would rather avoid - at least for the moment. I would also suggest that this function contains neither build nor solve calls directly. A more accurate name, by this logic, would simply be "fetch_strategic_model()"

Allow me some time to complete these changes. I have other priorities competing today.

Comment (tarnold17) [reply pareto/tests/test_utilities.py:41]: @Tominapa:

1) to be clear, I don't think you _need_ to restructure this code to use a fixture. Feel free to do so if you wish, but I don't think it's strictly necessary.

2) Your function `build_strategic_model()` calls `create_model()` and then `solve_model()` - these are the functions that build the strategic model and then solve it, respectively, so I'm not sure how you can argue that `build_strategic_model()` "contains neither build nor solve calls directly". I understand that the fact that you need to call `solve()` to get the quality model is an artifact of how the strategic model is designed. I'm just saying that it's easier to read code where function names describe as closely as possible the actions of the function (even if some of those actions are side effects). To be clear, this is a relatively minor thing, and I'm not going to block you from merging this PR or anything if you don't want to change the name of the function, I just want to explain my opinion clearly.

File: pareto/utilities/bounding_functions.py
Comment (nareshsusarla) [reply pareto/utilities/bounding_functions.py:58]: No, that is not correct, you need to adjust the value of the bound otherwise the value will be incorrect. Please review the two constraints in the model `PipelineCapacityExpansion` and `PipelineCapacity`. Although the index for pipeline is not modeled, but we it recognizes multiple pipelines through origin and destination and so, the capacity gets adjusted based on how many pipelines are there between 2 given nodes. An easiest way to do this, may be, is to add a factor of 2 to max() term in your method, i.e., 2 * max(..). A better way would be to identify the bi-directional pipelines and only apply the modification to them, your choice.

**Review by nareshsusarla — CHANGES_REQUESTED** [review]

File: pareto/tests/test_utilities.py
Comment (Tominapa) [reply pareto/tests/test_utilities.py]: Removed

File: pareto/tests/test_utilities.py
Comment (Tominapa) [reply pareto/tests/test_utilities.py]: 1. Done.
2. Indeed, the build and solve calls are in those other functions. I don't want to give the impression that I am recreating those capabilities here. I'm just calling an existing function to get a model on which to run tests. The name is now "fetch_strategic_model()" to be compatible with your guidelines.

File: pareto/utilities/bounding_functions.py
Comment (melody-shellman) [reply pareto/utilities/bounding_functions.py:61]: Good catch, Naresh!

**Review by melody-shellman — APPROVED** [review]
Good work here! I know  there will be a future iteration of bounding function, so I left a few notes in the description to help remind us what improvements we've already noted. Feel free to update with other items you think should be addressed for the next iteration.

**Review by tarnold17 — APPROVED** [review]

File: pareto/utilities/model_modifications.py
Comment (nareshsusarla) [inline pareto/utilities/model_modifications.py:40]:
```suggestion
                    index_var.unfix()
 ```

**Review by melody-shellman — APPROVED** [review]
Looks great! Good work !

**Comment by nareshsusarla** [comment]
looks good

**Pull request merged**
